src/object-store/src/layers/lru_cache.rs

```diff
@@ -120,120 +89,52 @@ impl<I: Accessor, C: Accessor> LayeredAccessor for LruCacheAccessor<I, C> {
     }
 
     async fn read(&self, path: &str, args: OpRead) -> Result<(RpRead, Self::Reader)> {
-        if !can_cache(path) {
-            return self.inner.read(path, args).await.map(to_output_reader);
-        }
-
-        let path = path.to_string();
-        let cache_path = self.cache_path(&path, &args);
-        let lru_cache = &self.lru_cache;
-
-        // the args is already in the cache path, so we must create a new OpRead.
-        match self.cache.read(&cache_path, OpRead::default()).await {
-            Ok((rp, r)) => {
-                increment_counter!(OBJECT_STORE_LRU_CACHE_HIT);
-
-                // update lru when cache hit
-                let mut lru_cache = lru_cache.lock().await;
-                let _ = lru_cache.get_or_insert(cache_path.clone(), || ());
-                Ok(to_output_reader((rp, r)))
-            }
-            Err(err) if err.kind() == ErrorKind::NotFound => {
-                increment_counter!(OBJECT_STORE_LRU_CACHE_MISS);
-
-                let (_, mut reader) = self.inner.read(&path, args.clone()).await?;
-                let (_, mut writer) = self.cache.write(&cache_path, OpWrite::new()).await?;
-
-                while let Some(bytes) = reader.next().await {
-                    writer.write(&bytes?).await?;
-                }
-
-                writer.close().await?;
-
-                match self.cache.read(&cache_path, OpRead::default()).await {
-                    Ok((rp, reader)) => {
-                        let r = {
-                            // push new cache file name to lru
-                            let mut lru_cache = lru_cache.lock().await;
-                            lru_cache.push(cache_path.clone(), ())
-                        };
-                        // delete the evicted cache file
-                        if let Some((k, _v)) = r {
-                            let _ = self.cache.delete(&k, OpDelete::new()).await;
-                        }
-                        return Ok(to_output_reader((rp, reader)));
-                    }
-                    Err(_) => return self.inner.read(&path, args).await.map(to_output_reader),
-                }
-            }
-            Err(err) => {
-                increment_counter!(OBJECT_STORE_LRU_CACHE_ERROR, OBJECT_STORE_LRU_CACHE_ERROR_KIND => format!("{}", err.kind()));
-                return self.inner.read(&path, args).await.map(to_output_reader);
-            }
-        }
+        self.read_cache
+            .read::<I, _>(path, args, |path, args| {
+                Box::pin(async move { self.inner.read(path, args.clone()).await })
+            })
```

Comment:
Can we clone `self.inner` and pass it to `read` instead of passing a future? Then we don't need so many lifetime parameters.

Reply:
I am afraid we can't. Because the inner was passed by `layer` function when implementing `Layer` trait. But the read cache is created before that.

Reply:
Looks like it is possible if we wrap `inner` with an `Arc`.
```suggestion
            .read(self.inner.clone(), path, args)
```